crates/sui-core/src/authority_active/checkpoint_driver/mod.rs

```diff
@@ -0,0 +1,878 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    sync::Arc,
+    time::Duration,
+};
+
+use parking_lot::Mutex;
+use sui_types::{
+    base_types::{AuthorityName, TransactionDigest},
+    error::SuiError,
+    messages::{CertifiedTransaction, ConfirmationTransaction, TransactionInfoRequest},
+    messages_checkpoint::{
+        AuthenticatedCheckpoint, AuthorityCheckpointInfo, CertifiedCheckpoint, CheckpointContents,
+        CheckpointFragment, CheckpointRequest, CheckpointResponse, CheckpointSequenceNumber,
+        SignedCheckpoint, SignedCheckpointProposal,
+    },
+};
+use tokio::time::timeout;
+
+use crate::{
+    authority_aggregator::{AuthorityAggregator, ReduceOutput},
+    authority_client::AuthorityAPI,
+    checkpoints::{proposal::CheckpointProposal, CheckpointStore},
+};
+use tracing::{debug, info, warn};
+use typed_store::Map;
+
+#[cfg(test)]
+pub(crate) mod tests;
+
+use super::ActiveAuthority;
+
+pub struct CheckpointProcessControl {
+    /// The time to allow upon quorum failure for sufficient
+    /// authorities to come online, to proceed with the checkpointing
+    /// main loop.
+    pub delay_on_quorum_failure: Duration,
+
+    /// The time between full iterations of the checkpointing
+    /// logic loop.
+    pub long_pause_between_checkpoints: Duration,
+
+    /// The time we allow until a quorum of responses
+    /// is received.
+    pub timeout_until_quorum: Duration,
+
+    /// The time we allow after a quorum is received for
+    /// additional responses to arrive.
+    pub extra_time_after_quorum: Duration,
+
+    /// The estimate of the consensus delay.
+    pub consensus_delay_estimate: Duration,
+
+    /// The amount of time we wait on any specific authority
+    /// per request (it could be byzantine)
+    pub per_other_authority_delay: Duration,
```

Comment:
Are all of those configuration parameters? If so, should we read them from NetworkConfig? (I am thinking about how we will hook this up with the rest of the validator)

Reply:
If not, we may just leave a TODO (I see you implement Default just below)

Reply:
mmh, how important is it for these parameters to be the same for all correct validators? It seems it is liveness-critical, should leave a clear note? (we may not want nodes operators playing with these)

Reply:
Checkpointing is an async protocol, so it should not do major damage if they are different across authorities.

Reply:
is it not partially-sync? the timeouts made me think so (I will check again)